### PR TITLE
Add a bunch of knife integration tests

### PR DESCRIPTION
**Pull request by thommay**
There's some bonus work to allow input on stdin to be passed to knife

### CODE REVIEW

**Comment by thommay** [comment]
🚧 still in progress 🚧

**Comment by thommay** [comment]
@chef/client-maintainers review appreciated; apologies for the enormitude.

File: spec/integration/knife/client_bulk_delete_spec.rb [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  include KnifeSupport
+
+  include_context "default config options"
+
+  when_the_chef_server "has some clients" do
+    before do
+      client "cons", {}
+      client "car", {}
+      client "cdr", {}
+      client "cat", {}
```
Comment (stevendanna) [inline spec/integration/knife/client_bulk_delete_spec.rb]: Would `concat` be a useful test case?

File: spec/integration/knife/client_bulk_delete_spec.rb [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    end
+
+    it "deletes all matching clients" do
+      knife("client bulk delete ^ca.*", input: "Y").should_succeed <<EOM
```
Comment (stevendanna) [inline spec/integration/knife/client_bulk_delete_spec.rb]: Reading this test, my primary question is "what happens if I don't specify the ^ anchor.

File: spec/integration/knife/client_create_spec.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    end
+
+    it "saves the private key to a file" do
+      Dir.mktmpdir do |tgt|
+        knife("client create -f #{tgt}/bah.pem bah").should_succeed stderr: out
+        expect(File).to exist("#{tgt}/bah.pem")
+      end
+    end
+
+    it "reads the public key from a file" do
```
Comment (stevendanna) [inline spec/integration/knife/client_create_spec.rb]: Can we test that the public key was actually sent up to the server?

**Comment by smurawski** [comment]
:+1:

File: spec/integration/knife/client_bulk_delete_spec.rb (hunk 2 above)
Comment (thommay) [inline spec/integration/knife/client_bulk_delete_spec.rb]: added a test for that case

File: spec/integration/knife/client_create_spec.rb (hunk 3 above)
Comment (thommay) [inline spec/integration/knife/client_create_spec.rb]: not easily - zero doesn't store the default key since it doesn't really care about it

**Pull request merged**